### ISSUE TITLE
Spire-agent fails to start; race condition on mounting bundle ConfigMap

### ISSUE THREAD

**Issue by richardstone** [spiffe/spire#2978]
* **Version**: 1.2.0; 1.2.3 (others not tested)
* **Platform**: Linux 5.13.0-40-generic#45 ~20.04.1-Ubuntu SMP x86_64
* **Subsystem**: spire-agent

Cluster setup:
4 worker nodes

Scenario:
Start the full spire stack (server and agents) at the same time.

I have just noticed that when multiple spire-agents trying to start and mount the bundle configmap at the same time, one of the agents restart right after deploy because of the following error:
`could not parse trust bundle: open /run/spire/bundle/bundle.crt: no such file or directory`

As I found out, the issue occures almost every time when I have the container images already downloaded before spire deploy, so that it is more likely to start every spire-agent at the same time. To me it looks like that k8s can not handle a single configmap to multiple containers at the same time (at least it is limited in some way), so it makes a race condition between spire-agents.

The failed agent can start up without an issue after an initial restart.

**Comment by azdagron** [comment]
>  k8s can not handle a single configmap to multiple containers at the same time

Is there any documentation supporting this? This would surprise me.

> The failed agent can start up without an issue after an initial restart.

I _think_ the more likely explanation is that when the entire SPIRE infrastructure is started at the same time, the SPIRE Agents can come up before the SPIRE Server. SPIRE Server is the entity which populates that ConfigMap via the `k8sbundle` notifier plugin. If SPIRE Agent containers come up either before the SPIRE Server has had a chance to populate that ConfigMap, or alternatively, before the worker node the SPIRE Agent is running on has had a chance to _observe the ConfigMap update and apply the diff to the mounted volume_, then the SPIRE Agent is going to observe an empty directory. One solution here is for you to employ an init container to wait for the bundle to be present in the volume before starting the agent. You can also just rely on agent restarts.

**Comment by richardstone** [comment]
> Is there any documentation supporting this? This would surprise me.

No, that's only my assumption.

I fully aggree with your thoughts. I do use an initContainer. Until this point I only waited until spire-server is reachable.
Now I tried to wait for the bundle to appear (mounted the same configmap into the init container). What I see now is that it takes very long time after a reachable spire-server, until the bundle reaches the init container.

Server is OK after 24 seconds
Bundle is OK after 78 seconds (from start)

After this, one of the agents restarted with the following:
`time="2022-04-26T09:01:30Z" level=error msg="Agent crashed" error="create attestation client: failed to dial dns:///spire-server:8081: context deadline exceeded: connection error: desc = \"transport: authentication handshake failed: x509svid: could not verify leaf certificate: x509: certificate signed by unknown authority (possibly because of \\\"x509: ECDSA verification failure\\\" while trying to verify candidate authority certificate \\\"SPIFFE\\\")\""`

Other spire-agents are running but logs are full with this:
`time="2022-04-26T09:01:41Z" level=error msg="No identity issued" method=FetchX509SVID pid=1426 registered=false service=WorkloadAPI subsystem_name=endpoints`

Spire-server log is full with it creating bundle for the k8s-workload-registrar (multiple times in a single second):
`time="2022-04-26T09:13:25Z" level=debug msg="Set bundle for object" external=false name=k8s-workload-registrar plugin_name=k8sbundle plugin_type=Notifier subsystem_name=catalog`

I'm trying to use Spire together with NSM, and NSM pods are restarting with them not getting x509 certificate.
`Apr 26 09:14:14.751 [FATA] error getting x509 source`


The only difference between a working scenario (with occasional initial spire-agent restart due to the initial issue described in the ticket) is that now I'm mounting the spire-bundle to the initContainer, and waiting for the bundle.cert (with _test -f_ command) to appear.
